main.tf

```diff
@@ -144,19 +144,43 @@ resource "aws_iam_instance_profile" "default" {
   role  = "${aws_iam_role.default.name}"
 }
 
-resource "aws_ecr_lifecycle_policy" "default" {
+resource "aws_ecr_lifecycle_policy" "tagged" {
+  count      = "${var.create_tagged_lifecycle == "true" ? 1 : 0}"
   repository = "${aws_ecr_repository.default.name}"
 
   policy = <<EOF
 {
   "rules": [{
     "rulePriority": 1,
-    "description": "Rotate images when reach ${var.max_image_count} images stored",
+    "description": "Rotate images when reach ${var.max_tagged_image_count} images stored",
     "selection": {
       "tagStatus": "tagged",
       "tagPrefixList": ["${var.stage}"],
       "countType": "imageCountMoreThan",
-      "countNumber": ${var.max_image_count}
+      "countNumber": ${var.max_tagged_image_count}
+    },
+    "action": {
+      "type": "expire"
+    }
+  }]
+}
+EOF
+}
+
+resource "aws_ecr_lifecycle_policy" "untagged" {
+  count      = "${var.create_untagged_lifecycle == "true" ? 1 : 0}"
+  repository = "${aws_ecr_repository.default.name}"
+
+  policy = <<EOF
+{
+  "rules": [{
+    "rulePriority": 2,
```

Comment:
I'd also like to bring attention to `rulePriority`. The AWS documentation says it's required, and that it must be unique for each lifecycle policy. I can't think of a scenario where an image would be affected by _both_ of these policies (an image is either tagged, or not), but maybe this has some implications for future lifecycle policies?

Reply:
Actually, I see now that `rules` is an array...maybe the tagged/untagged rules should both be in the same lifecycle policy?

Reply:
If each rule should be in its own policy (which makes it easier to enable only one or the other), then perhaps `rulePriority` here should remain set to `1`?

Reply:
@ivan-pinatti any opinion?

Reply:
Hi @osterman,

Yes, please refer to (https://github.com/cloudposse/terraform-aws-ecr/pull/7#issuecomment-361802899)[https://github.com/cloudposse/terraform-aws-ecr/pull/7#issuecomment-361802899].
We didn't implement the select option (ternary) because there was a restriction on Terraform side back then.

Also, the rule created was related to *image versions* not days.
Which I think should be adopted as a best practice for every scenario, I don't see any reason why to keep 100 Docker image versions for example.

Reply:
It seems safer that any setting that leads to a loss of artifacts should be disabled by default. Users who don't fully understand the module may be in for an unpleasant surprise if only a maximum of 7 tagged images are kept (currently the default for this module). 

A means of disabling any or all lifecycle policies gives more control to the end user, and reflects the full range of the options available in AWS (i.e. AWS does not require an ECR repository to have any lifecycle policies).

Reply:
I guess I could get rid of the `create_tagged_lifecycle` and `create_untagged_lifecycle` variables, and only attach the respective lifecycle policies if `max_tagged_image_count` or `max_untagged_image_days` is greater than 0. That would be a little less verbose.

Reply:
@justindisney seems like a good idea.

Reply:
You can set the max image count variable value to 1000 while Hashicorp doesn't provide a solution that would let us implement a ternary to make it optional.

As I said, thinking as a private Docker image repo in a deploy pipeline, I see no reason to keep more than 7 versions, (IMHO 3 is more than enough actually).
And even so, this is a safeguard to rollback in case your deployment doesn't work and that's it.

Anyhow, the lifecycle only rotates the image that has a specific tag prefix, if you don't want it, you can simply remove the tagging into your image. This was designed to keep the 7 latest versions of each environment (stg, qa, lt and prod for example). So, if you change to 1000 and consider these environments, you will have more than 4000 Docker images, which I think will be more than enough until they fix the issue, just saying...

However, this is just my opinion, it is your call @osterman. :)

AWS ECR limit docs - https://docs.aws.amazon.com/AmazonECR/latest/userguide/service_limits.html

Reply:
> , I see no reason to keep more than 7 versions, (IMHO 3 is more than enough actually).

That's totally unreasonable if the same repository is used for production and staging. An active CI/CD pipeline can generate lots of images in a day. That would have the effect of deleting the production copy. If you need to scale out production, now you're hosed. =)

Reply:
I think we should leave both disabled by default to @justindisney's point. I think adding a number of canned lifecycle policies the way you guys have done is great. Then let the user decide which strategy to employ. I don't think we have to satisfy both at the same time.

Reply:
@ivan-pinatti 
> Hashicorp doesn't provide a solution that would let us implement a ternary to make it optional

Can you jog my memory on why we can't use a ternary?

Reply:
@osterman it has a tag prefix for each environment, there is no risk of removing a PROD version when pushing a new version in STG.

It is 7 versions for PROD + 7 versions for STG in the same repo for example.
Look here [https://github.com/cloudposse/terraform-aws-ecr/blob/master/main.tf#L157](https://github.com/cloudposse/terraform-aws-ecr/blob/master/main.tf#L157).

The logical flow is always to increment and deploy a new version, you just go back __only__ when your deployment is blocked and a rollback to a previous version is the only possible way.
Even so, It is very unusual that you have to go back several versions.

Reply:
@osterman 

Furthermore, I'm totally in favor in making it optional to enable/disable. Actually I would like to have it done back then, I just didn't because of the restrictions.

But, between making the default with OR without the lifecycle, I'm totally to leave as-is until the issue is fixed.
For now, you can workaround by raising the variable value or not using the tag prefix.

Reply:
Even in staging, you might be running unlimited environments =) 

In our case, we'll run multiple staging environments inside of Kubernetes off of the same registry. Every branch could conceivably be a new environment. I think it's really hard to conceive of "sane" defaults. Thus, I think that @justindisney brings up a good point. The default should be to delete nothing and the caller can  choose. This is assuming it's valid to define multiple concurrent `aws_ecr_lifecycle_policy`

/begin Shameless plug!!!

Join us next Wednesday (3/14) @ 10:00am PST for a webinar+live demo on how to run "Unlimited Staging Environments" on Kubernetes using CodeFresh. https://codefresh.io/how-to-run-apps-on-kubernetes/

Reply:
@ivan-pinatti aha! this is the part you're talking about:
```
 "tagPrefixList": ["${var.stage}"]
```

Now that you call that to my attention, I think that might be overly opinionated =) It assumes a certain image tagging convention. @cloudposse/engineering does not follow this specific convention. We use a combination of branch, semver and git ref. 

I think that perhaps this should be a separate parameter. I suppose that if it's empty, it matches everything?

Reply:
@osterman 

That's the idea, you can have N environments but you will keep only the 7 latest versions in your Docker repository for each environment.

7 versions for DEV
7 versions for STG
7 versions for QA
7 versions for UAT
7 versions for LT
7 versions for PROD

You can deploy the same artifact (Docker image version) in several places as many times as you want.

Regarding the restriction, If I remember correctly we couldn't use the ternary to dynamic populate the rules properties for the lifecycle.

If

Reply:
@osterman if you go back to my PR #7 and my other PR in Jenkins repo [https://github.com/cloudposse/jenkins/pull/9](https://github.com/cloudposse/jenkins/pull/9), you will see that I went over this assuming tag as staging strategy, maybe I wouldn't was too emphatic on it, I'm sorry.

Reply:
@ivan-pinatti there is nothing wrong with this strategy... and I think it's great for `jenkins` and perhaps related projects, but enforcing it at *this* layer is not the best place. Let's just implement it in such a way that it supports both use-cases.

Reply:
@ivan-pinatti just for additional context, in our typical architectures, we have ~3 "stages" or (tiers) of service. Production Tier, Staging Tier and Dev Tier. Each one of these is a separate AWS organization. In the Staging tier, there would be one registry. Then perhaps 20 different environments (e.g. one for every open branch, including one for QA, UAT, etc). It's infeasible to provision infrastructure for every environment. But it is feasible to reuse the registry for all environments in that stage.